libopenage/screenshot.cpp

```diff
@@ -46,44 +52,89 @@ std::string ScreenshotManager::gen_next_filename() {
 
 
 void ScreenshotManager::save_screenshot() {
+	job::JobManager *job_mgr = this->job_manager;
+
+	uint32_t *pxdata = new uint32_t [this->window_size.x * this->window_size.y];
+	glReadPixels(0, 0,
+	             this->window_size.x, this->window_size.y,
+	             GL_RGBA, GL_UNSIGNED_BYTE, pxdata);
+
+	encode_png(pxdata);
```

Comment:
Depending if you want to transfer ownership, either do:

``` cpp
encode_png(std::move(pxdata));
```

to transfer the memory ownership (funcion arg needs to be a `std::unique_ptr<uint32_t[]> &&`

or

``` cpp
encode_png(pxdata.get());
```

The `.get()` returns an unowned raw-pointer to the allocated memory.

It may be better to transfer ownership and then write the data to the file in the `encode_png` function.

Alternative architectures would be that the `encode_png` function creates some memory chunk that is returned and can then be written to a file, or used as texture, etc.

Reply:
thanks. can you explain, why it needs a reference to a reference?

Reply:
That is called ["rvalue reference"](http://en.cppreference.com/w/cpp/language/reference). It prefers to bind to rvalues and extens the lifetime of a temporary object. The `move` is needed to make the `unique_ptr` an rvalue ("remove the name"), and we need to do that so the object is not copied to the function argument. Thus, the object is "transferred" into the function.